### ISSUE TITLE
"New" Mac OSX Docker Toolbox

### ISSUE THREAD

**Issue by asfaltboy** [wercker/support#131]
I'm having some trouble using wercker cli tool with a fresh installation of Docker Toolbox for OSX.

The toolbox seems to install and run the machine on a different (randomly generated?) host:port than what wercker expects. I found that I had to run `docker-machine ls` to find the url for my virtualbox host, e.g: `tcp://192.168.99.99:2376`.

However, even with the correct host, I'm still getting a fail when building locally:

    $ DOCKER_HOST=tcp://192.168.99.99:2376  wercker build
    ERROR Get https://192.168.99.99:2376/version: x509: certificate signed by unknown authority

Any idea what other steps I should perform to get started with wercker?

Note: I entered this scenario after reading [this deprecation message in docker docs](https://docs.docker.com/installation/mac/).

A [similar issue](https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/issues/1) in gitlab ci, though still on boot2docker, hints at a possible solutions. Here is my default docker machine connection config:

    $ docker-machine config default
    --tlsverify --tlscacert="/Users/big/.docker/machine/machines/default/ca.pem" --tlscert="/Users/big/.docker/machine/machines/default/cert.pem" --tlskey="/Users/big/.docker/machine/machines/default/key.pem" -H=tcp://192.168.99.99:2376

K, seems this "default" machine is borked, I started a new one as [described in the docs](https://docs.docker.com/v1.5/machine/):

    docker-machine create --driver virtualbox dev
    export DOCKER_TLS_VERIFY="1"
    export DOCKER_HOST="tcp://192.168.99.101:2376"
    export DOCKER_CERT_PATH="/Users/big/.docker/machine/machines/dev"
    export DOCKER_MACHINE_NAME="dev"
    # Run this command to configure your shell:
    # eval "$(docker-machine env dev)"

Now **with this** machine, and I can now run `DOCKER_HOST=tcp://192.168.99.101:2376  wercker build` and it works as intended.

This should probably be somehow added to the docs.

**Comment by hatchan** [comment]
The new docker machine method is not supported out of the box. You do need to specify the environment variables which docker client uses to connect to the docker daemon. See https://github.com/wercker/support/issues/119

**Comment by asfaltboy** [comment]
Thanks @hatchan, perhaps wercker can detect the correct host for "default" or "dev" machine before defaulting to localhost ?

**Comment by asfaltboy** [comment]
I believe I it should be simple to streamline the flow that works with `docker-machine`; if anyone needs it, this is what I do in my `.bashrc` (replace `default` with your machine name):

```bash
eval $(docker-machine env default)
export DOCKER_HOST=$(docker-machine url default)
```

**Comment by hatchan** [comment]
We are looking at integration with `docker-machine`. For the time being the above snippet is probably the best (though I don't think you need the second line, as `docker env` already includes `DOCKER_HOST`).

**Comment by asfaltboy** [comment]
good point thanks !

**Comment by mstade** [comment]
I've used the new Docker Toolbox for OSX for the past week, and have compiled these notes:

- `eval $(docker-machine env default)` never hurts
- If `wercker dev` or `wercker build` seems to hang; make sure the virtual machine is running by running `docker-machine status default` and if it's stopped, `docker-machine start default` will bring it back up
- If wercker fails with random errors after installing the Docker Toolbox, try removing the virtual machine and adding it again using the quickstart – I had to do this once, not sure what caused it
- Don't forget to `docker-machine stop` when you're done working, it's not lovely having VBoxHeadless hogging quite a bit of memory in the background
- If you're having troubles with symlinks (such as when running the `npm-install` step,) it's probably because VirtualBox v5.0.4 botched symlinks; fret not, a [fix is available][1]

I'm sure most of these would have showed up when using boot2docker as well. Overall, wercker seems to do what it says on the tin when working with the Docker Toolbox, however the latter is a bit of an annoying overhead in case you don't use docker much other than with wercker. It'd be nice if wercker could hide it away more, somehow. I've started looking at wercker as a viable alternative to Vagrant because of the `dev` pipeline – maybe being inspired by the Vagrant workflow wouldn't hurt?

[1]: https://www.virtualbox.org/ticket/14563

One thought would be for wercker to check for the status of a `default` machine, and if it's stopped then start it automagically – setting necessary environment variables as well – when running `wercker dev` or such. In so doing, wercker could let the user know that the VM is now running in the background, and if they want to stop it they'd have to run `wercker stop` or something similar. I don't know, I'm just shooting random thoughts, but it seems to me there might be enough competition in this space to provide inspiration for nice workflows.
